### PR TITLE
rgw: add abort multipart date and rule-id header to init multipart upload response

### PR DESCRIPTION
**Pull request by zhangsw**
If abort multipart upload is configured in lifecycle, `x-amz-abort-date` and `x-amz-abort-rule-id` should be returned in response header.
https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateMultipartUpload.html

Fixes: https://tracker.ceph.com/issues/45077

Signed-off-by: zhang Shaowen <zhangshaowen@cmss.chinamobile.com>

### CODE REVIEW

File: src/rgw/rgw_lc.cc [hunk 1]
```diff
@@ -1685,4 +1685,71 @@ std::string s3_expiration_header(
...
+                      << dendl;
+    return false;
+  } /* catch */
+
+  boost::optional<ceph::real_time> abort_date_tmp;
+  boost::optional<std::string> rule_id_tmp;
+  const auto& rule_map = config.get_rule_map();
+  for (const auto& ri : rule_map) {
+    const auto& rule = ri.second;
+    auto& id = rule.get_id();
```
Comment (yuvalif) [inline src/rgw/rgw_lc.cc]: all of these could be ``cosnt auto&``

File: src/rgw/rgw_lc.cc [hunk 2]
```diff
@@ -1685,4 +1685,71 @@ std::string s3_expiration_header(
...
+    auto& mp_expiration = rule.get_mp_expiration();
+    if (!rule.is_enabled()) {
+      continue;
+    }
+    if(!prefix.empty() && !boost::starts_with(obj_key.name, prefix)) {
+      continue;
+    }
+
+    boost::optional<ceph::real_time> rule_abort_date;
+    const LCExpiration& rule_abort_multipart =mp_expiration;
```
Comment (yuvalif) [inline src/rgw/rgw_lc.cc]: why do you need ``rule_abort_multipart``? can you just use ``mp_expiration``?

File: src/rgw/rgw_lc.cc [hunk 3]
```diff
@@ -1685,4 +1685,71 @@ std::string s3_expiration_header(
...
+  try {
+    config.decode(iter);
+  } catch (const buffer::error& e) {
+    ldpp_dout(dpp, 0) << __func__
+                      <<  "() decode life cycle config failed"
+                      << dendl;
+    return false;
+  } /* catch */
+
+  boost::optional<ceph::real_time> abort_date_tmp;
```
Comment (yuvalif) [inline src/rgw/rgw_lc.cc]: probably better to use ``std::optional`` in new code

File: src/rgw/rgw_lc.cc [hunk 4]
```diff
@@ -1685,4 +1685,71 @@ std::string s3_expiration_header(
...
+  boost::optional<std::string> rule_id_tmp;
```
Comment (yuvalif) [inline src/rgw/rgw_lc.cc]: probably better to use string_view here, so no copy is done in line 1748

**Comment by zhangsw** [comment]
@yuvalif I've updated this pr.

**Comment by cbodley** [comment]
jenkins test this please

**Comment by ivancich** [comment]
@zhangsw Would you add a tracker for this?

**Comment by yuvalif** [comment]
@zhangsw ping

**Comment by zhangsw** [comment]
@yuvalif I've added a tracker.

File: src/rgw/rgw_lc.cc [hunk 5]
```diff
@@ -1685,4 +1685,70 @@ std::string s3_expiration_header(
...
+      continue;
+    }
+    if(!prefix.empty() && !boost::starts_with(obj_key.name, prefix)) {
+      continue;
+    }
+
+    std::optional<ceph::real_time> rule_abort_date;
+    if (mp_expiration.has_days()) {
+      rule_abort_date = std::optional<ceph::real_time>(
+              mtime + make_timespan(mp_expiration.get_days()*24*60*60));
```
Comment (ofriedma) [inline src/rgw/rgw_lc.cc:1732]: lgtm

One small thing I'm not sure about is if the abort multipart date should be round up to midnight next day like expiration one
@zhangsw What do you think?

**Comment by ofriedma** [comment]
@zhangsw ping

**Comment by zhangsw** [comment]
@ofriedma You're right~ I've updated it.

**Comment by ofriedma** [comment]
Unstale

**Comment by ofriedma** [comment]
http://pulpito.ceph.com/ofriedma-2020-07-14_09:04:16-rgw-wip-zhangsw-expr-hdr-i01-distro-basic-smithi/

**Pull request merged**